### PR TITLE
fix(router): navigate to root if initial navigation fails

### LINKED ISSUE #16211: canActivate=>false can result in blank screen/dead link
**I'm submitting a ...**  (check one with "x")
```
[ X ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

canActivate, canActivateChild will result in NavigationCancel event if 'false'/reject() happens in the Guard. Additionally the URL is reset to the base path before the navigation attempt (as though to leave the user on the page they navigated one).

This current behavior means that if you just go directly to a location in your browser (e.g. localhost:4200/some-guarded-route) and it returns false the URL will read "localhost:4200/" with a blank page as navigation is canceled.

**Expected behavior**

The right behavior to me seems that the router should try to see if there are further matches in the routing table (i.e. the guard 'false' case would cover the use case scenario of #14515). By continuing down the route table the behavior would be to land on '**' route in the worst case while also enabling conditional routing.

Minimally NavigationCancel event should determine if the resetUrlToCurrentUrlTree is an already displayed route (either rendering it if necessary or doing nothing if it's already rendered).

I will see if I can issue a PR to fix this for review. Will need to determine what the side effects would be for cases where people would expect just to stay on the currently rendered page (considering it was already rendered). However, to me I think the correct behavior for a dead link should be to go down the route table and hit '**' for a 404 or something vs doing nothing when the user clicks on it and rendering nothing if the user hits the route directly before any route has been rendered.

**Minimal reproduction of the problem with instructions**

Plunker won't work for this because you would need to be able to hit the route as though it was bookmarked to reproduce.

Effectively just create project using CLI, create a route guard, return false in the methods. Then attempt to hit the route directly without loading a root non-guarded route first.

**What is the motivation / use case for changing the behavior?**

Motivation is that you won't have the user on a blank screen if they happen to hit a guarded route without meeting the conditions.

Second part is that by resolving these scenarios by continuing down the route table you would enable conditional routing for scenarios such as:

Creating routes like:
localhost:4200/:organization
localhost:4200/:user

If the route guard found a valid organization it could allow the request to move forward vs user. This is all gravy. A start is to not give the user a blank page when resetting the route via resetUrlToCurrentUrlTree when the current Url wasn't rendered in the first place.

**Please tell us about your environment:**

Mac OS Sierra (10.12.4), WebStorm, NPM, ng serve

* **Angular version:** 2.0.X

Angular 4.0.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Chrome 57.0.2987.133 (64-bit)
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.2.2

* **Node (for AoT issues):** `node --version` =   
Node 7.9

### PR DESCRIPTION
**Pull request by atscott**
When a navigation in the router fails, we reset the internal state and
browser URL to the last successful navigation. For the initial
navigation, this does not exist, so `''`/`'/'` is used. Prior to this
commit, the state reset was not also accompanied by a navigation to the
root of the application. This results in a "blank" screen (or rather, it
only shows content that is outside the `router-outlet` such as the
static app header).

With this commit, the router initializer, which includes an `initialNavigation`
call, now tracks whether the initial navigation succeeded (including
redirects). If the navigation did not succeed, a navigation to the root
of the application is triggered (`'/'`). Again, this is in addition to
what the router already does for this failed navigation: setting its internal
state and browser URL to the root.

Fixes #16211

### CODE REVIEW

**Comment by atscott** [comment]
[TGP](https://test.corp.google.com/OCL:458292348:BASE:458280241:1656618761022:2eca5e49)

File: packages/router/src/router_module.ts [hunk 1]
```diff
@@ -592,53 +633,25 @@ function provideEnabledBlockingInitialNavigation(): Provider {
...
-                initNavigation = true;
+              afterNextNavigation(router, () => {
+                function complete() {
+                  // Unblock APP_INITIALIZER in case the initial navigation was canceled or errored
+                  // without a redirect.
+                  resolve(true);
+                  initNavigation = true;
+                }
+                if (!router.navigated) {
+                  router.navigateByUrl('/').then(() => {
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:649]: Would it make sense to log a warning into a console (in ngDevMode) in this case (and in a similar case above) to notify developers that there was an extra navigation since initial one failed. Something like:
```
NG0XYZ: Initial navigation to `/some-rejected-route` failed, navigating to the `/` path instead.
```

File: packages/router/test/bootstrap.spec.ts [hunk 2]
```diff
@@ -307,6 +309,162 @@ describe('bootstrap', () => {
...
+        imports: [RouterModule],
+        standalone: true
+      })
+      class AppCmp {
+      }
+      const appRef = await bootstrapApplication(AppCmp, {
+        providers: [
+          importProvidersFrom(RouterModule.forRoot(
+              [
+                {path: '', component: HomeCmp},
```
Comment (AndrewKushnir) [inline packages/router/test/bootstrap.spec.ts:398]: I think it'd be great to capture a test case when navigating to `/` also fails - would it cause any issues?

**Review by AndrewKushnir — COMMENTED** [review]
@atscott looks great, just a couple comments on potential DX improvement and a corner case when navigation to `/` also fails.

File: packages/router/test/bootstrap.spec.ts (hunk 2 above)
Comment (atscott) [reply packages/router/test/bootstrap.spec.ts:398]: It would result in the same blank screen. Navigating to `/` _should not_ ever fail. I would consider this a problem with the application configuration.

Comment (atscott) [reply packages/router/test/bootstrap.spec.ts:398]: There's nothing really to test here from the `Router`. This behavior is essentially "undefined" if it's rejected.

Comment (AndrewKushnir) [reply packages/router/test/bootstrap.spec.ts:398]: Ok, got it. I just wanted to make sure there is no infinite loop in this case.

File: packages/router/src/router_module.ts (hunk 1 above)
Comment (sonukapoor) [reply packages/router/src/router_module.ts:649]: Can we also add a message to the router trace as well when tracing is enabled in the router module. Could be also done in a followup PR.

Comment (atscott) [reply packages/router/src/router_module.ts:649]: The router trace will already contain the `NavigationStart` event so I don't think an additional one is necessary. I can add the warning in the console.

Comment (AndrewKushnir) [reply packages/router/src/router_module.ts:649]: > I can add the warning in the console.

Yeah, having a console warn would be great 👍

File: packages/router/src/router_module.ts [hunk 3]
```diff
@@ -658,6 +673,10 @@ function provideEnabledBlockingInitialNavigation(): Provider {
   ];
 }
 
+function warnOfFailedInitialNavigation() {
+  console.warn('Initial navigation failed. Navigating to application root.');
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:677]: Do we have info on which route we've tried to navigate to? I think it'd be great to include that into the warning as well:
```suggestion
  console.warn('Initial navigation to `/path/a/b/c` failed. Navigating to application root.');
```

**Review by AndrewKushnir — APPROVED** [review]

File: packages/router/src/router_module.ts (hunk 3 above)
Comment (atscott) [reply packages/router/src/router_module.ts:677]: The initial navigation may have included several navigation attempts because it includes redirects. Tracking the attempted router navigations is possible but would require quite a bit of code. The other option would be to depend on the internal implementation of `router.initialNavigation` and initial navigation attempt is `inject(location).path(true)`. 

The second option would mean saving the path before the `router.initialNavigationCall` and then passing it to `warnOfFailedInitialNavigation`
```
const initialNavigationUrl = NG_DEV_MODE ? inject(Location).path(true) : '';
```

**Review by sonukapoor — APPROVED** [review]

File: packages/router/src/router_module.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/router/src/router_module.ts:677]: Discussed offline, there are no extra changes to the error message planned in this PR.

**Comment by atscott** [comment]
Closing - something like this is probably sufficient:

```
bootstrapApplication(App, {}).then(async (app) => {
  const router = app.injector.get(Router);
  await app.whenStable();
  if (!router.lastSuccessfulNavigation) {
    router.navigateByUrl('/');
  }
});
```
